Source/JavaScriptCore/dfg/DFGMovHintRemovalPhase.cpp

```diff
@@ -0,0 +1,155 @@
+#include "OperandsInlines.h"
+
+namespace JSC { namespace DFG {
+
+namespace {
+
+namespace DFGMovHintRemovalPhaseInternal {
+static constexpr bool verbose = false;
+}
+
+class MovHintRemovalPhase : public Phase {
+public:
+    MovHintRemovalPhase(Graph& graph)
+        : Phase(graph, "MovHint removal")
+        , m_insertionSet(graph)
+        , m_state(OperandsLike, graph.block(0)->variablesAtHead)
+        , m_changed(false)
+    {
+    }
+
+    bool run()
+    {
+        if (DFGMovHintRemovalPhaseInternal::verbose) {
+            dataLog("Graph before MovHint removal:\n");
+            m_graph.dump();
+        }
+
+        for (BasicBlock* block : m_graph.blocksInNaturalOrder())
+            handleBlock(block);
+
+        m_insertionSet.execute(m_graph.block(0));
+
+        return m_changed;
+    }
+
+private:
+    void handleBlock(BasicBlock* block)
+    {
+        if (DFGMovHintRemovalPhaseInternal::verbose)
+            dataLog("Handing block ", pointerDump(block), "\n");
+
+        // A MovHint is unnecessary if the local dies before it is used. We answer this question by
+        // maintaining the current exit epoch, and associating an epoch with each local. When a
+        // local dies, it gets the current exit epoch. If a MovHint occurs in the same epoch as its
+        // local, then it means there was no exit between the local's death and the MovHint - i.e.
+        // the MovHint is unnecessary.
+
+        Epoch currentEpoch = Epoch::first();
+
+        m_state.fill(Epoch());
+        m_graph.forAllLiveInBytecode(
+            block->terminal()->origin.forExit,
+            [&] (Operand reg) {
+                m_state.operand(reg) = currentEpoch;
+            });
+
+        if (DFGMovHintRemovalPhaseInternal::verbose)
+            dataLog("    Locals at ", block->terminal()->origin.forExit, ": ", m_state, "\n");
+
+        // Assume that blocks after us exit.
+        currentEpoch.bump();
+
+        for (unsigned nodeIndex = block->size(); nodeIndex--;) {
+            Node* node = block->at(nodeIndex);
+
+            if (node->op() == MovHint) {
+                Epoch localEpoch = m_state.operand(node->unlinkedOperand());
+                if (DFGMovHintRemovalPhaseInternal::verbose)
+                    dataLog("    At ", node, " (", node->unlinkedOperand(), "): current = ", currentEpoch, ", local = ", localEpoch, "\n");
+                if (!localEpoch || localEpoch == currentEpoch) {
+                    node->setOpAndDefaultFlags(ZombieHint);
+                    UseKind useKind = node->child1().useKind();
+                    Node* constant = m_constants.ensure(static_cast<std::underlying_type_t<UseKind>>(useKind), [&]() -> Node* {
+                        return m_insertionSet.insertBottomConstantForUse(0, m_graph.block(0)->at(0)->origin, useKind).node();
+                    }).iterator->value;
+                    node->child1() = Edge(constant, useKind);
+                    m_changed = true;
```

Comment:
Don't we have to mark everything up to the next Epoch as not exitOK here? What stops a later phase from inserting an exiting instruction in the middle? e.g.

```
@1 = ...
@2 = MovHint(@1, loc1)
// Insert ForceOSR exit here.
@3 = ArithAdd(@2, exitOK)
```
If we turn @2 into a ZombieHint won't loc1 now be the bottom value when executing the add instruction in LLInt/Baseline?

Reply:
If we insert it in the middle, then I think it is still fine since we insert this ZombieHint based o. bytecode's liveness. So regardless of exit happens, this local will not be accessed after the exit (so any value is fine).

Reply:
That exitOK thing does not work since Exit can be inserted so long as exit does not use the dead locals.
So, I think that we are running this mov-hint-removal phase at very end since we keep the invariant we do not make locals live again, so I think this patch is fine.

Reply:
I think if we're going to have that invariant it should be documented both here and probably in DFGPlan.cpp.

Reply:
Sounds good! I've added it to the DFGPlan.cpp etc.